src/ORM/DataList.php

```diff
@@ -287,15 +287,19 @@ public function canFilterBy($fieldName)
     /**
      * Return a new DataList instance with the records returned in this query
      * restricted by a limit clause.
-     *
-     * @param int $limit
-     * @param int $offset
-     * @return static
      */
-    public function limit($limit, $offset = 0)
+    public function limit(?int $length, int $offset = 0): static
     {
-        return $this->alterDataQuery(function (DataQuery $query) use ($limit, $offset) {
-            $query->limit($limit, $offset);
+        if ($length !== null && $length < 0) {
+            throw new InvalidArgumentException('$length can not be a negative value');
+        }
+
+        if ($offset < 0) {
+            throw new InvalidArgumentException('$offset can not be a negative value');
+        }
+
```

Comment:
Exceptions would be thrown any way by the queries below eventually. Seems better to just do it now.